go/engine/saltpack_user_keyfinder.go

```diff
@@ -0,0 +1,282 @@
+func (e *SaltpackUserKeyfinder) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *SaltpackUserKeyfinder) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func (e *SaltpackUserKeyfinder) GetPublicKIDs() []keybase1.KID {
+	var r []keybase1.KID
+	for _, keys := range e.RecipientDeviceAndPaperKeyMap {
+		r = append(r, keys...)
+	}
+	for _, keys := range e.RecipientEntityKeyMap {
+		r = append(r, keys...)
+	}
+
+	return r
+}
+
+func (e *SaltpackUserKeyfinder) GetSymmetricKeys() []libkb.SaltpackReceiverSymmetricKey {
+	return []libkb.SaltpackReceiverSymmetricKey{}
+}
+
+func (e *SaltpackUserKeyfinder) Run(m libkb.MetaContext) (err error) {
+	defer m.CTrace("SaltpackUserKeyfinder#Run", func() error { return err })()
+
+	if len(e.Arg.TeamRecipients) != 0 {
+		m.CDebugf("tried to use SaltpackUserKeyfinder for a team. This should never happen")
+		return fmt.Errorf("cannot find keys for teams")
+	}
+
+	err = e.AddOwnKeysIfNeeded(m)
+	if err != nil {
+		return err
+	}
+
+	err = e.lookupAndAddRecipients(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *SaltpackUserKeyfinder) AddOwnKeysIfNeeded(m libkb.MetaContext) error {
+	if !e.Arg.NoSelfEncrypt {
+		if !m.ActiveDevice().Valid() {
+			return libkb.NewLoginRequiredError("need to be logged in or use --no-self-encrypt")
+		}
+		arg := libkb.NewLoadUserArgWithMetaContext(m).WithUID(m.ActiveDevice().UID()).WithForcePoll(true)
+		upak, _, err := m.G().GetUPAKLoader().LoadV2(arg)
+		if err != nil {
+			return err
+		}
+		e.AddUserRecipient(m, &upak.Current)
+	}
+	return nil
+}
+
+// lookupAndAddRecipients adds the KID corresponding to each recipient to the recipientMap
+func (e *SaltpackUserKeyfinder) lookupAndAddRecipients(m libkb.MetaContext) error {
+	for _, u := range e.Arg.Recipients {
+		// TODO make these lookups in parallel (maybe using sync.WaitGroup)
+		upk, err := e.LookupUser(m, u) // For existing users
+		if err != nil {
+			return fmt.Errorf("Cannot find keys for %v: it is not an assertion for a registered user (err = %v)", u, err)
+		}
+		err = e.AddUserRecipient(m, upk)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *SaltpackUserKeyfinder) LookupUser(m libkb.MetaContext, user string) (upk *keybase1.UserPlusKeysV2, err error) {
+
+	Arg := keybase1.Identify2Arg{
+		UserAssertion: user,
+		Reason: keybase1.IdentifyReason{
+			Type: keybase1.IdentifyReasonType_ENCRYPT,
+		},
+		AlwaysBlock:      true,
+		IdentifyBehavior: keybase1.TLFIdentifyBehavior_CLI,
+	}
+	eng := NewResolveThenIdentify2(m.G(), &Arg)
+	if err := RunEngine2(m, eng); err != nil {
+		return nil, libkb.IdentifyFailedError{Assertion: user, Reason: err.Error()}
+	}
+
+	engRes := eng.Result()
+	if engRes == nil {
+		return nil, fmt.Errorf("Null result from Identify2")
+	}
+	arg := libkb.NewLoadUserArgWithMetaContext(m).WithUID(engRes.Upk.GetUID()).WithForcePoll(true)
```

Comment:
there's a race possible here, which is that the user reset their account between the identify and the UPAK load.  It's too bad we don't return the upak from ID2 engine, but we don't. It's probably the right idea to check for the race and then error out if you hit it.  See `engine.ResolveAndCheck`